Support converting column to a python list in pylibcudf

Matt711 opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->
Adds an API for converting a `Column` to a python list.
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

Matt711 committed:
Support converting column to a python list in pylibcudf

Matt711 committed:
lint

Matt711 commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1028,6 +1028,39 @@ cdef class Column:
+        """
+        cdef unique_ptr[scalar] s
+
+        if not plc_is_fixed_width(self.type()) and self.type().id() != type_id.STRING:
```
For fixed width non-null columns, we should probably do the copy in one batch via RMM.

Matt711 committed:
Merge branch 'branch-25.10' into fea/plc/to-python-list

Matt711 committed:
Merge branch 'branch-25.10' into fea/plc/to-python-list

Matt711 committed:
Use arrow C data interface and python c api

Matt711 committed:
include dlpack

Matt711 committed:
Merge branch 'branch-25.10' into fea/plc/to-python-list

Matt711 committed:
remove include

Matt711 committed:
combine _arrow_to_pylist impls

Matt711 committed:
remove interop import from column

Matt711 committed:
merge conflict

Matt711 committed:
Merge branch 'branch-25.10' into fea/plc/to-python-list

Matt711 committed:
clean up

TomAugspurger commented on python/pylibcudf/tests/test_column_from_iterable.py:
```diff
@@ -216,3 +216,36 @@ def test_from_nested_list_of_large_strings(dummy_large_string_type):
+        ([1.25, 4.5, 7.75], plc.DataType(plc.TypeId.FLOAT32)),
+        ([1.25, 4.5, 7.75], plc.DataType(plc.TypeId.FLOAT64)),
+        ([True, False, True], plc.DataType(plc.TypeId.BOOL8)),
+        (["foo", "bar", "baz"], plc.DataType(plc.TypeId.STRING)),
```
Could you add a test for that includes some chars that are multiple bytes when utf-8 encoded? Just since we're indexing into the buffer, it'd be good to have a test case that exercises that a bit.

And would it be possible to also parametrize this by some null handling? Maybe a new fixture that parametrizes over `[True, False]` for whether to set the first item to null?

I also see some special treatment of length-0 columns. It'd be good to test that too (here, or in another test; whatever is easiest).

Finally, you've probably gotten the bitmask stuff correct, but even if you parametrize this test by whether there are any nulls, we'd still just be dealing with the first byte of the bitmask (I think). Maybe a separate test with some longer data?

Matt711 committed:
Merge branch 'branch-25.10' into fea/plc/to-python-list

Matt711 committed:
address review

Matt711 reviewed (commented):
Going to let at least @vyasr review again before merging

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+
+    cdef const char* base = <const char*>buffers[1] + item_size * <size_t>offset
+
+    if null_bm is NULL:
```
I see what you're trying to optimize here, but until we have a use case where we see the performance impact of checking the bitmask nullity N times instead of once I'm inclined to say duplicating the entire if/else cascade for types isn't worth it.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+
+    for i in range(n):
+        bit_index = <size_t>(offset + <int64_t>i)
+        if not _is_valid(null_bm, bit_index):
```
Unlike in the numeric case, this function isn't handling the case where `null_bm` is NULL.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+
+    for i in range(n):
+        bit_index = <size_t>(offset + <int64_t>i)
+        if not _is_valid(null_bm, bit_index):
+            _set_item(out, <Py_ssize_t>i, None)
+            continue
+
+        val = ((data_bm[bit_index >> 3] >> (bit_index & 7)) & 1) != 0
```
Alternatively, I think that this logic could be put into `_arrow_from_pylist_numeric` so that you don't need a completely separate function here.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+    return out
+
+
+cdef list _arrow_to_pylist_string(ArrowArray* arr):
```
This function does have some more differences from the numeric/bool cases, but I still think it's similar enough to be worth combining them. You'll need to maintain a few extra variables for every invocation on non-numeric data (like the offsets) and those will just be null for those types, but the rest of the logic can just be shoved into an else clause for string data.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1319,3 +1347,263 @@ def is_c_contiguous(
+
+    if null_bm is NULL:
+        for i in range(n):
+            if dtype == type_id.INT8:
```
After you apply my other suggestions around trying to combine the numeric and bool overloads, it might also be worthwhile to try and rewrite the whole function as inline C++ instead of using Cython. The reason that I say that is that I suspect that the casts below aren't doing anything other than making Cython happy (they aren't actually affecting correctness) and the C/C++ will handle automatic castig for you, meaning that a lot of these cases might collapse.

Matt711 replied:
Let me know what you think of the cython refactor. It was pretty easy to include the string case in with the bool and numeric cases.

vyasr replied:
The code is definitely much improved now, but I think it is worth giving my suggestion a shot still. I suspect it would make `_arrow_to_pylist_impl` much simpler without a lot of casts, you could more easily localize some variable declarations without Cython's restrictions on where cdef statements are valid, and you could probably avoid exposing more ArrowArray properties to our Cython. Plus you could write switch statements rather than if/else cascades (Cython should optimize those if it can, but then we're relying on its heuristics rather than much more well-defined C/C++ compiler behaviors for a switch statement).

Matt711 replied:
Thanks I'll give this a go. Do you think I should do this in a follow-up? I'm asking because it'll take me some more time to translate to C++ and we're in burndown soon. Although this is a low priority PR, so maybe it's okay to take the time to work out whether we want to do this in C++ in this PR. 

IIUC the overall structure should be something like the following. I'm skipping specifying a header because I'm not sure how to make the cmake changes yet.
```cython
cdef extern from *:
    """
    ...
    """
    PyObject* cpp_arrow_to_pylist(int dtype, ArrowArray* arr)
```

For the implementation, I am thinking a template to build the python object for the types that use the same C python call (eg. https://github.com/rapidsai/cudf/pull/19446/files#diff-a3c2e9bea60260df24118e4d2f6c63c402e52b23cefe414310ab326fc2e6cde4R1436-R1440 and https://github.com/rapidsai/cudf/pull/19446/files#diff-a3c2e9bea60260df24118e4d2f6c63c402e52b23cefe414310ab326fc2e6cde4R1443-R1458).

And then another template for the main loop which should build the python list. Maybe the type parameter can be a callable that takes the array index `i` and the `bit_index` and returns the python object at that index? It should be a template because the implementation differs from bool to fixed width, to string. Is this plan on the right track or would you go with a different approach?

vyasr commented on python/pylibcudf/pylibcudf/libcudf/interop.pxd:
```diff
@@ -21,15 +21,34 @@ cdef extern from "dlpack/dlpack.h" nogil:
 
 # The Arrow structs are not namespaced.
 cdef extern from "cudf/interop.hpp" nogil:
+    # https://arrow.apache.org/docs/format/CDataInterface.html#structure-definitions
```
Rather than adding the entire declaration for completeness, let's just declare the parts that we actually need to access from Cython (which I think is almost none of the schema attributes, but probably almost all of the array attributes?).

Matt711 committed:
address review

Matt711 committed:
remove unused import

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+    # 3. Shift the target bit to the least significant bit position
+    #    0b10101101 >> 6 = 0b00000010
+    # 4. Remove all other bits
+    #    0b00000010 & 1 = 0 (=> the 7th element is null)
```
I'm sort of confused by what this very long example is meant to show. Are you trying to explain how the arrow bitmask is set up? If so, would linking to https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps be sufficient?

Matt711 replied:
> Are you trying to explain how the arrow bitmask is set up?

Yes, I originally added to help myself understand how the bitmap worked. But I think the docs do a good-enough job (especially the formula for `_is_valid`)

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+
+
+cdef inline void _set_item(list out, int64_t i, object obj) except *:
+    # https://docs.python.org/3/c-api/list.html#c.PyList_SetItem
```
```suggestion
    # PyList_SetItem steals a reference, so this helper function
    # ensures that we incref the object before calling it
    # https://docs.python.org/3/c-api/list.html#c.PyList_SetItem
```

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
+
+    if dtype in (
+        type_id.INT8, type_id.INT16, type_id.INT32, type_id.INT64,
+        type_id.UINT8, type_id.UINT16, type_id.UINT32, type_id.UINT64,
+        type_id.FLOAT32, type_id.FLOAT64, type_id.BOOL8, type_id.STRING
+    ):
```
I would invert the condition and throw the exception. That way the rest of the function can be deindented. Also for this many values it might be more efficient to use a set than a tuple.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+    cdef bint is_string = dtype == type_id.STRING
+    cdef bint check_valid = arr.null_count != 0
+
+    if arr.n_buffers >= 1 and buffers[0] != NULL and arr.null_count != 0:
+        null_bm = <const uint8_t*>buffers[0]
```
Are the first two clauses in this condition necessary? My thoughts:
- Can you ever have n_buffers == 1? I think you must always have at least two to hold the bitmask and the data, right? Let me know if I'm missing something there.
- If `buffers[0] == NULL` then `null_bm = buffers[0]` is a no-op since you start off with it as NULL anyway.

If I'm right, you either assign `null_bm` unconditionally but continue preempting the call below with `check_valid` (and this condition just becomes `cdef const uint8_t* null_bm = <const uint8_t*>buffers[0] if check_valid else NULL`).

Matt711 replied:
I think you have it right. See
- https://arrow.apache.org/docs/format/Columnar.html#fixed-size-primitive-layout
- https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout

The physical memory layout of bool is fixed size.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+        type_id.UINT8, type_id.UINT16, type_id.UINT32, type_id.UINT64,
+        type_id.FLOAT32, type_id.FLOAT64, type_id.BOOL8, type_id.STRING
+    ):
+        if is_string:
+            if arr.n_buffers < 3 or buffers[1] == NULL or buffers[2] == NULL:
+                for i in range(N):
+                    _set_item(out, i, None)
+                return out
+        else:
+            if arr.n_buffers < 2 or buffers[1] == NULL:
+                for i in range(N):
+                    _set_item(out, i, None)
+                return out
```
```suggestion
        # 
        if (
            (is_string and (arr.n_buffers < 3 or buffers[1] == NULL or buffers[2] == NULL))
            or (arr.n_buffers < 2 or buffers[1] == NULL)
        ):
            for i in range(N):
                _set_item(out, i, None)
            return out
```

vyasr replied:
Probably should also add comments on when you would hit this case.

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1354,3 +1393,181 @@ def is_c_contiguous(
+            numeric_base = <const char*>buffers[1] + item_size * <size_t>offset
+
+        for i in range(N):
+            bit_index = offset + <int64_t>i
```
Minor nit, you don't need the bit index for non-nullable string/integral types. You'd have to duplicate the line twice below but it might be worth saving the computation.

Matt711 replied:
Done

vyasr reviewed (changes requested):
Looks much better! I think we're close here. Still some room to improve though I think.

Matt711 committed:
address review

Matt711 committed:
address review

Matt711 committed:
remove cast

Matt711 committed:
Merge branch 'branch-25.12' into fea/plc/to-python-list

vyasr commented on python/pylibcudf/pylibcudf/column.pyx:
```diff
@@ -1401,3 +1424,219 @@ def is_c_contiguous(
+            // ensures that we incref borrowed singletons before calling it.
+            size_t bit_index = size_t(offset + i);
+            if (check_valid && !_is_valid(null_bm, bit_index)) {
+                Py_INCREF(Py_None);
```
See my comment below, also immortal starting in 3.12.

vyasr commented on python/pylibcudf/pylibcudf/libcudf/interop.pxd:
I think all of the changes in this file are unnecessary now.

vyasr reviewed (approved):
Right now the changes to use inline C++ look more verbose, but I think with my suggestion it should get shorter and also produce more efficient code than the original Cython implementation.

vyasr reviewed (changes requested):
Sorry meant to request changes.

vyasr committed:
Address PR review feedback (non-controversial items)

- Revert ArrowArray field declarations in interop.pxd (inline C++ accesses
  fields directly via nanoarrow headers, Cython doesn't need them)
- Replace into_py<T> struct with if constexpr in numeric_converter
- Make check_valid a template parameter in _make_pylist to eliminate
  per-element branch via if constexpr
- Use auto for type declarations in inline C++
- Add Python 3.12 immortality TODO comments for Py_INCREF on singletons
- Switch PyList_SET_ITEM to PyList_SetItem (stable ABI compatible)
- Add _set_item helper for borrowed singleton incref+set pattern

vyasr committed:
Merge upstream/main into fea/plc/to-python-list

Resolve conflicts in column.pxd, column.pyx, and libcudf/interop.pxd.
Fix to_pylist to pass stream to to_arrow_host_raw after upstream API change.
Remove duplicate from_iterable_of_py in column.pyi.

vyasr committed:
Use type_dispatcher with custom IdTypeMap for to_pylist

Replace the manual switch/if-else dispatch in _dispatch_pylist with
cudf::type_dispatcher and a custom IdTypeMap (dispatch_pylist_type).

The IdTypeMap maps arithmetic types and cudf::string_view to themselves,
and all unsupported type_ids (timestamps, decimals, lists, structs, etc.)
to void — collapsing them into a single trivial template instantiation
that returns nullptr.

A pylist_functor<check_valid> struct handles the three supported paths
via if constexpr: bool (bit-packed), string_view (offsets + char data),
and numeric (fixed-width). This follows the established libcudf pattern
(cf. dispatch_void_if_non_integral in reductions/bitwise.cu).

vyasr committed:
Merge remote-tracking branch 'upstream/main' into fea/plc/to-python-list

vyasr committed:
Style

vyasr committed:
Merge branch 'main' into fea/plc/to-python-list

vyasr committed:
Handle large string offsets (int64) in to_pylist

The string overload previously hard-coded int32_t for Arrow offset
buffers. However, cudf STRING columns can have INT64 offsets (large
strings exceeding 2GB), and to_arrow_host produces int64_t offsets
for these columns.

Add a large_offsets flag threaded from the Python layer (detected via
the column's offsets child type) through to the C++ functor. The string
SFINAE overload now branches on this flag to cast buffers[1] as either
int64_t* or int32_t*.

Add a pa.large_string() test case to exercise this path.

vyasr committed:
Revert unnecessary changes

vyasr committed:
Remove superfluous comments

Matt711 committed:
Merge branch 'main' into fea/plc/to-python-list
